src/adapter/src/coord/timeline.rs

```diff
@@ -481,65 +490,81 @@ impl Coordinator {
         global_timelines.get_mut(timeline).expect("inserted above")
     }
 
-    pub(crate) fn remove_storage_ids_from_timeline<I>(&mut self, ids: I) -> Vec<Timeline>
-    where
-        I: IntoIterator<Item = GlobalId>,
-    {
-        let mut empty_timelines = Vec::new();
-        for id in ids {
-            if let TimelineContext::TimelineDependent(timeline) = self.get_timeline_context(id) {
-                let TimelineState { read_holds, .. } = self
-                    .global_timelines
-                    .get_mut(&timeline)
-                    .expect("all timelines have a timestamp oracle");
-                read_holds.remove_storage_id(&id);
-                if read_holds.is_empty() {
-                    self.global_timelines.remove(&timeline);
-                    empty_timelines.push(timeline);
-                }
+    /// Groups together storage and compute resources into a [`CollectionIdBundle`]
+    pub(crate) fn build_collection_id_bundle(
+        &self,
+        storage_ids: impl IntoIterator<Item = GlobalId>,
+        compute_ids: impl IntoIterator<Item = (ComputeInstanceId, GlobalId)>,
+        clusters: impl IntoIterator<Item = ComputeInstanceId>,
+    ) -> CollectionIdBundle {
+        let mut compute: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
+
+        // Collect all compute_ids.
+        for (instance_id, id) in compute_ids {
+            compute.entry(instance_id).or_default().insert(id);
+        }
+
+        // Collect all GlobalIds associated with a compute instance ID.
+        let cluster_set: BTreeSet<_> = clusters.into_iter().collect();
+        for (_timeline, TimelineState { read_holds, .. }) in &self.global_timelines {
+            let holds = read_holds
+                .compute_ids()
+                .filter(|(instance_id, _id)| cluster_set.contains(instance_id));
+            for (instance_id, ids) in holds {
+                let ids = ids.map(|(_antichain, id)| id);
+                compute.entry(*instance_id).or_default().extend(ids);
+            }
+        }
+
+        CollectionIdBundle {
+            storage_ids: storage_ids.into_iter().collect(),
+            compute_ids: compute,
+        }
+    }
+
+    pub(crate) fn remove_resources_associated_with_timeline(
```

Comment:
Can you add a doc-comment that explains what the return type means.

Reply:
Done!